exploration/text-vs-code.md

```diff
@@ -0,0 +1,224 @@
+
+ICU MessageFormat and Fluent both support inline selectors
+separated from the text using `{…}` for multi-variant messages.
+ICU MessageFormat is the only known format that uses `{…}` to also delimit text.
+
+[Mustache templates](https://mustache.github.io/mustache.5.html)
+and related languages wrap "code" in `{{…}}`.
+In addition to placeholders that are replaced by their interpolated value during formatting,
+this also includes conditional blocks using `{{#…}}`/`{{/…}}` wrappers.
+
+[Handlebars](https://handlebarsjs.com/guide/) extends Mustache expressions
+with operators such as `{{#if …}}` and `{{#each …}}`,
+as well as custom formatting functions that become available as e.g. `{{bold …}}`.
+
+[Jinja templates](https://jinja.palletsprojects.com/en/3.1.x/templates/) separate
+`{% statements %}` and `{{ expressions }}` from the base text.
+The former may define tests that determine the inclusion of subsequent text blocks in the output.
+
+A cost that the message formatting and templating languages mentioned above need to rely on
+is some rule or behaviour that governs how to deal with whitespace at the beginning and end of a pattern,
+as statements may be separated from each other by newlines or other constructs for legibility.
+
+Other formats supporting multiple message variants tend to rely on a surrounding resource format to define variants,
+such as [Rails internationalization](https://guides.rubyonrails.org/i18n.html#pluralization) in Ruby or YAML
+and [Android String Resources](https://developer.android.com/guide/topics/resources/string-resource.html#Plurals) in XML.
+These formats rely on the resource format providing clear delineation of the beginning and end of a pattern.
+
+Based on available data,
+no more than 0.3% of all messages and no more than 0.1% of messages with variants
+contain leading or trailing whitespace.
+No more than one third of this whitespace is localizable,
+and most commonly it's due to improper segmentation or other internationalization bugs.
+
+## Use-Cases
+
+Most messages in any localization system do not contain any expressions, statements or variants.
+These should be expressible as easily as possible.
+
+Many messages include expressions that are meant to be replaced during formatting.
+For example, a greeting like "Hello, {$username}!" would be formatted with the variable
+`$username` being replaced by an input variable.
+
+In some rare cases, replacement variables might be added (or removed) in one particular
+locale versus messages in other locales.
+
+Sometimes, the replacement variables need to be formatted.
+For example, formatting a message like `You have {$distance} kilometers to go`
+requires that the numeric value `{$distance}` be formatted as a number according to the locale: `You have 1,234 kilometers to go`.
+
+Formatting of replacement variables might also require tailoring.
+For example, if the user wants to show fractions of a kilometer in the above example
```

Comment:
Nit: This use of "user" has a different meaning than the use in «_a greeting like "Hello, {$username}!"_» three paragraphs earlier. We haven't been great about vocabulary here, but I think it would make sense to explicitly reference the [Web Platform Priority of Constituencies](https://www.w3.org/TR/design-principles/#priority-of-constituencies) ("user" vs. "author" vs. "implementor" vs. "specification writer" vs. "theoretical purity") or define and use analogous terms (e.g., "developer" and "translator" rather than "author").

Reply:
I've refactored the text to refer to "authors", "developers" and/or "translators" rather than "users".

I think bringing in the priority of constituencies is beyond the scope of this design doc, but it could be valuable to introduce separately.